src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract bool IsWrittenTo(TIdentifierNameSyntax node, SemanticModel model, CancellationToken cancellationToken);
+        protected abstract bool IsMemberOfThisInstance(SyntaxNode node);
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected void AnalyzeType(SyntaxNodeAnalysisContext context)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferReadonly, context.Node.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var typeSymbol = (ITypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node);
+
+            var nonReadonlyFieldMembers = PooledHashSet<IFieldSymbol>.GetInstance();
+            foreach (var item in typeSymbol.GetMembers())
+            {
+                if (item is IFieldSymbol symbol &&
+                    symbol.DeclaredAccessibility == Accessibility.Private &&
+                    !symbol.IsReadOnly &&
+                    !symbol.IsConst &&
+                    !symbol.IsImplicitlyDeclared &&
+                    !IsMutableValueType(symbol.Type))
+                {
+                    var isInCurrentNode = false;
+                    foreach (var syntaxReference in item.DeclaringSyntaxReferences)
+                    {
+                        if (!syntaxReference.SyntaxTree.Equals(context.Node.SyntaxTree))
+                        {
+                            continue;
+                        }
+
+                        if (!context.Node.Span.Contains(syntaxReference.Span))
+                        {
+                            continue;
+                        }
+
+                        isInCurrentNode = true;
+                        break;
+                    }
+
+                    if (isInCurrentNode)
+                    {
+                        nonReadonlyFieldMembers.Add(symbol);
+                    }
+                }
+            }
+
+            var syntaxFactsService = GetSyntaxFactsService();
+            foreach (var syntaxReference in typeSymbol.DeclaringSyntaxReferences)
```

Comment:
>typeSymbol.DeclaringSyntaxReferences [](start = 44, length = 36)

Note that compiler does not cache declaring syntax references on the symbol, and this will allocate an array every time. Consider adding code prior to this loop to check for empty `nonReadonlyFieldMembers` and bail out.